File: python/leetcode/diagonal-traverse.py
class Solution:
    def findDiagonalOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        up = True # current direction of traversal
        y = 0
        x = 0
        result = []
        while y < len(matrix) and x < len(matrix[0]):
            result.append(matrix[y][x])
            if up:
                if x + 1 < len(matrix[0]) and y - 1 >= 0:
                    x = x + 1
                    y = y - 1
                elif x + 1 >= len(matrix[0]):
                    up = False
                    y = y + 1
                elif y - 1 < 0:
                    up = False
                    x = x + 1
            elif not up:
                if y + 1 < len(matrix) and x - 1 >= 0:
                    x = x - 1
                    y = y + 1
                elif y + 1 >= len(matrix):
                    up = True
                    x = x + 1
                elif x - 1 < 0:
                    up = True
                    y = y + 1
        return result

File: java/leetcode/diagonal-traverse.java
class Solution {
    public int[] findDiagonalOrder(int[][] matrix) {
        if (matrix.length == 0) {
            int[] result = new int[0];
            return result;
        }
        boolean up = true;
        int y = 0;
        int x = 0;
        int[] result = new int[matrix.length * matrix[0].length];
        int i = 0;
        while (y < matrix.length && x < matrix[0].length) {
          result[i] = matrix[y][x];
          i++;
          if (up) {
            if (x + 1 < matrix[0].length && y - 1 >= 0) {
              x++;
              y--;
            } else if (x + 1 >= matrix[0].length) {
              up = false;
              y++;
            } else if (y - 1 < 0) {
              up = false;
              x++;
            }
          } else {
            if (y + 1 < matrix.length && x - 1 >= 0) {
              x--;
              y++;
            } else if (y + 1 >= matrix.length) {
              up = true;
              x++;
            } else if (x - 1 < 0) {
              up = true;
              y++;
            }
          }
        }
        return result;
    }
}
